### PR TITLE
[MAUT-7878] Export Custom Item

### PR DESCRIPTION
**Pull request by priyanshu2316**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N]
| New feature/enhancement? (use the a.x branch)      | [ Y]
| Deprecations?                          | [ N]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ Y] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | https://backlog.acquia.com/browse/MAUT-7878

#### Description:

This PR is to export customItems with linkedContactIds.

#### Steps to test this PR:
1. create a custom object with any number of fields.
2. create any number of custom items for newly created custom object in step 1. 
3. can link any number of custom items to contacts.(can link more than 1 contacts to any customItem).
4. beside new button in custom item list page , will view a button for export as csv, click on that.
5. Can check the email will receive a link to a file in zip format, click on the file, it will download the file. can view the contents and check with data we created in steps 1-3.

note:
1. linkedContactIds will be comma separated, if the custom item is linked to multiple contact ids or will be empty if no contact id is linked. please check https://github.com/acquia/mc-cs-plugin-custom-objects/wiki/Import for format of csv file.

### CODE REVIEW

File: Command/CustomItemsScheduledExportCommand.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        $this->formatterHelper             = $formatterHelper;
+
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->setName(self::COMMAND_NAME)
+            ->setDescription('Export customItems which are scheduled in `customItem_export_scheduler` table.')
```
Comment (tejasync) [inline Command/CustomItemsScheduledExportCommand.php]:
```suggestion
            ->setDescription('Export customItems which are scheduled in `custom_item_export_scheduler` table.')
```

File: Config/config.php [hunk 2]
```diff
@@ -10,7 +10,7 @@
 $coParams = [
     'name'        => 'Custom Objects',
     'description' => 'Adds custom objects and fields features to Mautic',
-    'version'     => '0.0.19',
+    'version'     => '0.0.23',
```
Comment (tejasync) [inline Config/config.php:13]: Q. Is this version is related to migration version?

File: Controller/CustomItem/ExportController.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+use Symfony\Component\HttpFoundation\File\Exception\FileNotFoundException;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Response;
+
+class ExportController extends AbstractFormController
+{
+    /**
+     * @var CustomItemPermissionProvider
+     */
+    private $permissionProvider;
```
Comment (tejasync) [inline Controller/CustomItem/ExportController.php]:
```suggestion
    private CustomItemPermissionProvider $permissionProvider;
```

File: Controller/CustomItem/ExportController.php [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+     * @param CustomItemPermissionProvider $permissionProvider
+     * @param CustomItemExportSchedulerModel $model
+     */
+    public function __construct(
+        CustomItemPermissionProvider $permissionProvider,
+        CustomItemExportSchedulerModel $model
+    )
+    {
+        $this->permissionProvider = $permissionProvider;
+        $this->model = $model;
```
Comment (tejasync) [inline Controller/CustomItem/ExportController.php]:
```suggestion
        $this->permissionProvider = $permissionProvider;
        $this->model              = $model;
```

File: Entity/CustomItemExportScheduler.php [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+use MauticPlugin\CustomObjectsBundle\Repository\CustomItemExportSchedulerRepository;
+use Symfony\Component\Validator\Constraints as Assert;
+use Symfony\Component\Validator\Mapping\ClassMetadata as ValidatorClassMetadata;
+
+class CustomItemExportScheduler
+{
+    /**
+     * @var int
+     */
+    private int $id;
```
Comment (tejasync) [inline Entity/CustomItemExportScheduler.php]:
```suggestion
    private int $id;
```

File: Entity/CustomItemExportScheduler.php [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * @var User
+     */
+    private User $user; // Created by
```
Comment (tejasync) [inline Entity/CustomItemExportScheduler.php]:
```suggestion
    private User $user; // Created by
```

File: Entity/CustomItemExportScheduler.php [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * @var DateTimeImmutable
+     */
+    private DateTimeImmutable $scheduledDateTime;
```
Comment (tejasync) [inline Entity/CustomItemExportScheduler.php]:
```suggestion
    private DateTimeImmutable $scheduledDateTime;
```

File: Entity/CustomItemExportScheduler.php [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * @var int
+     */
+    private int $customObjectId;
```
Comment (tejasync) [inline Entity/CustomItemExportScheduler.php]:
```suggestion
    private int $customObjectId;
```

File: Migrations/Version_0_0_22.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+use Doctrine\DBAL\Schema\Schema;
+use Mautic\IntegrationsBundle\Migration\AbstractMigration;
+
+class Version_0_0_22 extends AbstractMigration
+{
+    /**
+     * @var string
+     */
+    private $table = 'custom_item_export_scheduler';
```
Comment (tejasync) [inline Migrations/Version_0_0_22.php]:
```suggestion
    private string $table = 'custom_item_export_scheduler';
```

File: Migrations/Version_0_0_23.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function up(): void
+    {
+        $customItemExportSchedulerTableName = $this->concatPrefix($this->table);
+
+        $this->addSql("ALTER TABLE ".$customItemExportSchedulerTableName." ADD scheduled_datetime DATETIME NOT NULL COMMENT '(DC2Type:datetime_immutable)'");
```
Comment (tejasync) [inline Migrations/Version_0_0_23.php]: Since this has not been released to production yet, can we merge this migration with above?

File: Model/CustomItemExportSchedulerModel.php [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+use MauticPlugin\CustomObjectsBundle\Repository\CustomItemRepository;
+use MauticPlugin\CustomObjectsBundle\Repository\CustomItemXrefContactRepository;
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
+use Symfony\Component\HttpFoundation\Response;
+
+class CustomItemExportSchedulerModel extends AbstractCommonModel
+{
+    private const EXPORT_FILE_NAME_DATE_FORMAT = 'Y_m_d_H_i_s';
+    private const CUSTOM_ITEM_LIMIT = 200;
+    private const CONTACT_LIMIT = 1000000;
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
    private const EXPORT_FILE_NAME_DATE_FORMAT = 'Y_m_d_H_i_s';
    private const CUSTOM_ITEM_LIMIT            = 200;
    private const CONTACT_LIMIT                = 1000000;
```

File: Model/CustomItemExportSchedulerModel.php [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+
+    /**
+     * @var ExportHelper
+     */
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
```

File: Model/CustomItemExportSchedulerModel.php [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+    private ExportHelper $exportHelper;
+
+    /**
+     * @var MailHelper
+     */
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
```

File: Model/CustomItemExportSchedulerModel.php [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+    private MailHelper $mailHelper;
+
+    /**
+     * @var CustomItemRouteProvider
+     */
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
```

File: Model/CustomItemExportSchedulerModel.php [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+    private CustomItemRouteProvider $customItemRouteProvider;
+
+    /**
+     * @var CustomFieldValueModel
+     */
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
```

File: Model/CustomItemExportSchedulerModel.php [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+    private CustomFieldValueModel $customFieldValueModel;
+
+    /**
+     * @var CustomItemXrefContactRepository
+     */
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
```

File: Model/CustomItemExportSchedulerModel.php [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+    private CustomItemXrefContactRepository $customItemXrefContactRepository;
+
+    /**
+     * @var CustomItemRepository
+     */
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
```

File: Model/CustomItemExportSchedulerModel.php [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+    private CustomItemRepository $customItemRepository;
+
+    /**
+     * @var string
+     */
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
```

File: Model/CustomItemExportSchedulerModel.php [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+
+    public function __construct(
+        ExportHelper     $exportHelper,
+        MailHelper       $mailHelper,
+        CustomFieldValueModel $customFieldValueModel,
+        CustomItemRouteProvider $customItemRouteProvider,
+        CustomItemXrefContactRepository $customItemXrefContactRepository,
+        CustomItemRepository $customItemRepository
+    )
+    {
```
Comment (tejasync) [inline Model/CustomItemExportSchedulerModel.php]:
```suggestion
    ) {
```

File: Repository/CustomItemExportSchedulerRepository.php [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+
+declare(strict_types=1);
+
+namespace MauticPlugin\CustomObjectsBundle\Repository;
+
+use Mautic\CoreBundle\Entity\CommonRepository;
+
+class CustomItemExportSchedulerRepository extends CommonRepository
+{
+}
```
Comment (tejasync) [inline Repository/CustomItemExportSchedulerRepository.php]:
```suggestion
}

```

File: Tests/Unit/Command/CustomItemExportCommandTest.php [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+use Symfony\Component\Console\Tester\CommandTester;
+
+class CustomItemExportCommandTest extends CustomObjectTestCase
+{
+    protected $useCleanupRollback = false;
+
+    /**
+     * @var CustomItemModel
+     */
+    private $customItemModel;
```
Comment (tejasync) [inline Tests/Unit/Command/CustomItemExportCommandTest.php]:
```suggestion
    private CustomItemModel $customItemModel;
```

File: Tests/Unit/Command/CustomItemExportCommandTest.php [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+    private string $customItemExportSchedulerIds;
+
+    /**
+     * @var CustomObject
+     */
```
Comment (tejasync) [inline Tests/Unit/Command/CustomItemExportCommandTest.php]:
```suggestion
```

File: Tests/Unit/Controller/CustomItem/ExportControllerTest.php [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+use MauticPlugin\CustomObjectsBundle\Entity\CustomObject;
+use MauticPlugin\CustomObjectsBundle\Tests\Unit\CustomObjectTestCase;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+class ExportControllerTest extends CustomObjectTestCase
+{
+    /**
+     * @var CustomObject
+     */
```
Comment (tejasync) [inline Tests/Unit/Controller/CustomItem/ExportControllerTest.php]:
```suggestion
```

File: Translations/en_US/flashes.ini [hunk 24]
```diff
@@ -3,4 +3,5 @@ custom.item.error.items.not.found="Items %ids% were not found"
 custom.item.error.items.denied="The action has been denied for items %ids%"
 custom.item.unlinked="%itemName% (%itemId%) was unlinked from %entityType% %entityId%"
 custom.item.linked="%itemName% (%itemId%) was linked to %entityType% %entityId%"
-custom.object.error.used.in.segments="Custom object "%name%" (#%id%) cannot be deleted because it's used in the following Segment(s): %segments%."
\ No newline at end of file
+custom.object.error.used.in.segments="Custom object "%name%" (#%id%) cannot be deleted because it's used in the following Segment(s): %segments%."
+custom.item.export.being.prepared="File is getting processed. The file will be send to your email address '%user_email%' You can download the file from the link provided in the email."
```
Comment (tejasync) [inline Translations/en_US/flashes.ini]:
```suggestion
custom.item.export.being.prepared="File is getting processed. The file will be send to your email address '%user_email%' You can download the file from the link provided in the email."

```

File: Command/CustomItemsScheduledExportCommand.php (hunk 1 above)
Comment (priyanshu2316) [reply Command/CustomItemsScheduledExportCommand.php]: done

File: Config/config.php (hunk 2 above)
Comment (priyanshu2316) [reply Config/config.php:13]: yes, it is related to migration versions.

File: Controller/CustomItem/ExportController.php (hunk 4 above)
Comment (priyanshu2316) [reply Controller/CustomItem/ExportController.php]: done

File: Migrations/Version_0_0_23.php (hunk 10 above)
Comment (priyanshu2316) [reply Migrations/Version_0_0_23.php]: yeah but doing that will not run migrations on dev. have tried that before. so thats why created new migration for it.

**Review by tejasync — CHANGES_REQUESTED** [review]
Hi @priyanshu2316, as discussed yesterday, please "cs-fix" all files from this PR.

**Comment by priyanshu2316** [comment]
@ts-navghane . Ran cs fixer for all files in the PR.

**Review by tejasync — DISMISSED** [review]

**Comment by rohitp19** [comment]
@priyanshu2316 Can you please resolve the conflicts?

**Comment by priyanshu2316** [comment]
@rohitp19 resolved the conflicts

File: CustomItemEvents.php [hunk 25]
```diff
@@ -140,4 +140,30 @@ final class CustomItemEvents
      * @var string
      */
     public const ON_CAMPAIGN_TRIGGER_CONDITION = 'custom.item.on_campaign_trigger_notification';
+
+    /**
+     * The custom.item.on_custom_item_export event is fired when the custom item is being exported.
+     *
+     * The event listener receives a
+     *
```
Comment (jazeabby) [inline CustomItemEvents.php]: Can we remove incomplete comments?

File: CustomItemEvents.php [hunk 26]
```diff
@@ -140,4 +140,30 @@ final class CustomItemEvents
...
+     * @see MauticPlugin\CustomObjectsBundle\Event\CustomItemExportSchedulerEvent
+     *
+     * @var string
+     */
+    public const ON_CUSTOM_ITEM_SCHEDULE_EXPORT = 'custom.item.on_custom_item_export_scheduled';
+
+    /**
+     * The custom.item.custom_item_prepare_export_file event is fired when the custom item data is being prepared to be exported.
+     *
+     * The event listener receives a
```
Comment (jazeabby) [inline CustomItemEvents.php]: Here too

File: Entity/CustomItemExportScheduler.php [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+
+    public function getCustomObject(): int
+    {
+        return $this->customObjectId;
+    }
+
+    /**
+     * @return $this
+     */
+    public function setCustomObjectId(int $customObjectId)
```
Comment (jazeabby) [inline Entity/CustomItemExportScheduler.php]:
```suggestion
    public function setCustomObjectId(int $customObjectId): self
```

File: Translations/en_US/messages.ini [hunk 28]
```diff
@@ -17,6 +17,7 @@ custom.object.relationship.master_object.label="Master Object"
 custom.object.tab.fields="Fields"
 custom.object.noresults.tip="Check the search filters or create your first Custom Object."
 custom.object.delete.confirm="Are you sure you want to delete this custom object?"
+custom.object.export.notblank = "Custom object id is required."
```
Comment (jazeabby) [inline Translations/en_US/messages.ini]:
```suggestion
custom.object.export.notblank="Custom object id is required."
```

**Review by jazeabby — COMMENTED** [review]
@priyanshu2316 Please look at the suggested changes

File: Migrations/Version_0_0_23.php [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+
+use Doctrine\DBAL\Schema\Schema;
+use Mautic\IntegrationsBundle\Migration\AbstractMigration;
+
+class Version_0_0_23 extends AbstractMigration
+{
+    /**
+     * @var string
+     */
+    private $table = 'custom_item_export_scheduler';
```
Comment (jazeabby) [inline Migrations/Version_0_0_23.php]:
```suggestion
    private string $table = 'custom_item_export_scheduler';
```

File: CustomItemEvents.php (hunk 25 above)
Comment (priyanshu2316) [reply CustomItemEvents.php]: done

**Comment by rohitp19** [comment]
@priyanshu2316 I can see related mc-cs PR is marked as WIP also I have checked exporting custom object item which didn't worked as no job gets added to customitems tube Can you please check and do the needful?
Thanks

File: Model/CustomItemExportSchedulerModel.php [hunk 30]
```diff
@@ -0,0 +1,271 @@
...
+use MauticPlugin\CustomObjectsBundle\Repository\CustomItemRepository;
+use MauticPlugin\CustomObjectsBundle\Repository\CustomItemXrefContactRepository;
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
+use Symfony\Component\HttpFoundation\Response;
+
+class CustomItemExportSchedulerModel extends AbstractCommonModel
+{
+    private const EXPORT_FILE_NAME_DATE_FORMAT = 'Y_m_d_H_i_s';
+    private const CUSTOM_ITEM_LIMIT            = 200;
+    private const CONTACT_LIMIT                = 1000000;
```
Comment (rohitp19) [inline Model/CustomItemExportSchedulerModel.php]: Also can you please confirm if 1M comma separated values get  stored in single cell. IMO it needs to be low.

**Comment by priyanshu2316** [comment]
@rohitp19 yes, the jobs wont be queued now as we have planned for MAUT-8093 to go first right? thats why we made MAUT-7878 to WIP so MAUT-7878 and MAUT-8093 don't get conflicts.

**Comment by rohitp19** [comment]
> @rohitp19 yes, the jobs wont be queued now as we have planned for MAUT-8093 to go first right? thats why we made MAUT-7878 to WIP so MAUT-7878 and MAUT-8093 don't get conflicts.

In this scenario we need to remove this from sprint Can you please raise the same in todays scrum call?

**Comment by priyanshu2316** [comment]
sure I will do that

**Comment by priyanshu2316** [comment]
@rohitp19 @jazeabby  @ts-navghane  I was testing with 50K contact ids to prepare csv file, the csv file is getting prepared but it will not be imported in sheets(tested with importing in google sheet) so in there on importing it did not load anything for contact ids. I also checked one cell has a size of 32k bytes size in csv file. So will update the PR with suitable values. Also wanted to ask in scenario where custom items is more than 1M , and 1 csv file might not have 1M or more rows. So for such scenario do we need to divide data in more than 1 file and send that many links to user.

**Comment by rohitp19** [comment]
@priyanshu2316 The way we did at other places of export if we have records more than 1 million we are not splitting files but add records in same csv file we need to keep the same behaviour here too.

**Review by tejasync — DISMISSED** [review]
LGTM! Thanks @priyanshu2316

**Review by jazeabby — DISMISSED** [review]
LGTM

File: Command/CustomItemsScheduledExportCommand.php [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $ids                        = $this->formatterHelper->simpleCsvToArray($input->getOption('ids'), 'int');
+        $customItemExportSchedulers = $this->customItemExportSchedulerModel->getRepository()->findBy(['id' => $ids]);
+        $count                      = 0;
+
+        foreach ($customItemExportSchedulers as $customItemExportScheduler) {
+            $customItemExportSchedulerEvent = new CustomItemExportSchedulerEvent($customItemExportScheduler);
+            $this->logger->info('starting exporting custom items for customItemExportScheduler id: '.$customItemExportScheduler->getId());
```
Comment (tejasync) [inline Command/CustomItemsScheduledExportCommand.php]: Logging with id won't be helpful as this row would get deleted from the `mautic_custom_item_export_scheduler` table.

Comment (priyanshu2316) [reply Command/CustomItemsScheduledExportCommand.php]: added audit log as we discussed.

File: Model/CustomItemExportSchedulerModel.php [hunk 32]
```diff
@@ -199,7 +205,7 @@ private function addCustomItemsToCsvFile(array $customFields, CustomObject $cust
 
                     $rowData   = $savedRow;
                     $rowData[] = implode(',', $results);
-                    $this->exportHelper->echoTouchJob();
+                    $this->eventDispatcher->dispatch(CustomItemEvents::ON_PROCESSING_FILE);
```
Comment (jazeabby) [inline Model/CustomItemExportSchedulerModel.php]: I think as this is used in allydebundle, we should add this check as well
`if ($this->eventDispatcher->hasListeners(CustomItemEvents::ON_PROCESSING_FILE)) {`

**Review by jazeabby — COMMENTED** [review]
the changes looks good, can you look at the mentioned comment? thanks

File: Model/CustomItemExportSchedulerModel.php (hunk 32 above)
Comment (priyanshu2316) [reply Model/CustomItemExportSchedulerModel.php]: thats a good point.

**Review by tejasync — DISMISSED** [review]

**Review by jazeabby — DISMISSED** [review]
Thanks

**Review by rohitp19 — APPROVED** [review]

**Review by escopecz — APPROVED** [review]
